treadmill/infra/instances.py

```diff
@@ -0,0 +1,234 @@
+        )
+
+    def upsert_dns_record(self, hosted_zone_id, domain='', Reverse=False):
+        self._change_resource_record_sets(
+            'UPSERT',
+            hosted_zone_id,
+            domain,
+            Reverse
+        )
+
+    def delete_dns_record(self, hosted_zone_id, domain='', Reverse=False):
+        self._change_resource_record_sets(
+            'DELETE',
+            hosted_zone_id,
+            domain,
+            Reverse
+        )
+
+    def _change_resource_record_sets(
+            self,
+            action,
+            hosted_zone_id,
+            domain='',
+            Reverse=False
+    ):
+        forward_dns_name = self.name.lower() + '.' + domain + '.'
+        _hosted_zone_id = hosted_zone_id.split('/')[-1]
+        _name, _type, _value = [
+            self._reverse_dns_record_name(),
+            'PTR',
+            forward_dns_name
+        ] if Reverse else [
+            forward_dns_name,
+            'A',
+            self.private_ip
+        ]
+        _conn = Connection('route53')
+        _conn.change_resource_record_sets(
+            HostedZoneId=_hosted_zone_id,
+            ChangeBatch={
+                'Changes': [{
+                    'Action': action,
+                    'ResourceRecordSet': {
+                        'Name': _name,
+                        'Type': _type,
+                        'TTL': 3600,
+                        'ResourceRecords': [{
+                            'Value': _value
+                        }]
+                    }
+                }]
+            }
+        )
+
+    def _reverse_dns_record_name(self):
+        ip_octets = self.private_ip.split('.')
+        ip_octets.reverse()
+        ip_octets.append('in-addr.arpa')
+
+        return '.'.join(ip_octets)
+
+
+class Instances:
+    def __init__(self, instances):
+        self.instances = instances
+        self.volume_ids = []
+        self.conn = Connection()
+
+    @property
+    def ids(self):
+        return [i.id for i in self.instances]
+
+    @classmethod
+    def load_json(cls, ids=[], filters=[]):
+        """Fetch instance details"""
+        conn = Connection()
+        response = []
+
+        if ids:
+            response = conn.describe_instances(
+                InstanceIds=ids
+            )['Reservations']
+        elif filters:
+            response = conn.describe_instances(
+                Filters=filters
+            )['Reservations']
+
+        return sum([r['Instances'] for r in response], [])
+
+    @classmethod
+    def get(cls, ids=[], filters=[]):
+        json = Instances.load_json(ids=ids, filters=filters)
+        return Instances(
+            instances=[Instance(
+                id=j['InstanceId'],
+                metadata=j
+            ) for j in json]
+        )
+
+    @classmethod
+    def create(cls, Name=None, ImageId=None,
+               InstanceType='t2.small', SubnetId='',
+               Count=1, SecurityGroupIds=None, KeyName='ms_treadmill_dev',
```

Comment:
not sure what ms_treadmill_dev means... in any case, it seems like a wrong default to me. instance type - do not pass default. i think it is best to use defaults like these in CLI, not in implementation.